### PR TITLE
chore: upgrade to yarn 2 (#5279)

### PR DESCRIPTION
**Pull request by nsinghal12**
This PR contains ui code for yarn upgrade to yarn 2

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
Looks like we'll have to disable the Snyk integration to get this to work. https://github.com/snyk/cli/issues/1518

**Comment by crenshaw-dev** [comment]
Not sure what this means:

> #39 [argocd-ui 6/6] RUN HOST_ARCH=amd64 NODE_ENV='production' NODE_ONLINE_ENV='online' NODE_OPTIONS=--max_old_space_size=8192 yarn build
> #39 1.258 Usage Error: Couldn't find the node_modules state file - running an install might help (findPackageLocation)

File: test/e2e/fixture/fixture.go [hunk 1]
```diff
@@ -782,6 +782,7 @@ func CreateSubmoduleRepos(repoType string) {
 	FailOnErr(Run(submoduleParentDirectory(), "chmod", "777", "."))
 	FailOnErr(Run(submoduleParentDirectory(), "git", "init"))
 	FailOnErr(Run(submoduleParentDirectory(), "git", "add", "."))
+	FailOnErr(Run(submoduleParentDirectory(), "git", "config", "--global", "protocol.file.allow", "always"))
```
Comment (crenshaw-dev) [inline test/e2e/fixture/fixture.go:785]:
```suggestion
```

This has been removed in master.  :-)

Comment (nsinghal12) [reply test/e2e/fixture/fixture.go:785]: Oh ok - missed it. Thanks.

**Comment by nsinghal12** [comment]
Hi @crenshaw-dev I tried multiple things today to try fix the `uuid/v1` error:

1. Adding `uuid/v1` as a dev dependency
2. Adding `uuid` as a `packageExtension` in `.yarnrc.yml`
3. Try allowing `yarn` to update lock file
4. Adding `yarn set version berry` and `yarn install --immutable` to step on installing dependencies

But the code is unable to resolve the package. 

As I see `uuid/v1` is not a package in NPM registry. It is the `uuid` package from which `v1.js` is being imported. I believe we will need to modify the `argo-ui` source code to switch from CommonJS `require` to ES `import` statements. (refer https://github.com/argoproj/argo-ui/blob/master/src/components/form-field/form-field.tsx#L7)

I haven't yet found guidance in my Google search on how to resolve CJS level-2 dependencies. Any advice on where to look, or how to resolve without modifying `argo-ui` ?

**Comment by crenshaw-dev** [comment]
@keithchong I think this PR is very close to being done. Can you take a look at the build failures? I don't know yarn well enough to have any ideas.

**Comment by keithchong** [comment]
Hi @crenshaw-dev, I haven't actually migrated to yarn 2 before.  Did you follow the steps here? https://yarnpkg.com/getting-started/migration#step-by-step

**Comment by blakepettersson** [comment]
> I believe we will need to modify the argo-ui source code to switch from CommonJS require to ES import statements. (refer https://github.com/argoproj/argo-ui/blob/master/src/components/form-field/form-field.tsx#L7)

It was a while since I've worked with yarn, but IIRC yarn 2 requires all dependencies to be explicitly be stated in `package.json` and from what I can see [it is not](https://github.com/argoproj/argo-ui/blob/master/package.json#L24-L42).

It could be a good opportunity to modify argo-ui to ES import, and to perhaps upgrade [uuid](https://github.com/uuidjs/uuid) to a newer version?

**Comment by blakepettersson** [comment]
@nsinghal12 @keithchong @crenshaw-dev I think the required changes are in argoproj/argo-ui#367.

**Comment by blakepettersson** [comment]
@ nsinghal12 if you have time you can try to merge `master` and see if it builds now.

**Comment by nsinghal12** [comment]
> @nsinghal12 if you have time you can try to merge `master` and see if it builds now.

Hello @blakepettersson I merged master and updated this branch but looks like build is failing at install node dependencies in integration tests.

**Comment by blakepettersson** [comment]
@nsinghal12 I think your `yaml.lock` file is malformed (it seems to contain Git conflict diffs). I'd suggest removing it and then regenerating the lockfile.

**Comment by nsinghal12** [comment]
> @nsinghal12 I think your `yaml.lock` file is malformed (it seems to contain Git conflict diffs). I'd suggest removing it and then regenerating the lockfile.

I think you meant `yarn.lock` file. I deleted and regenrated the file again and pushed it. But looks like same issue.

**Comment by crenshaw-dev** [comment]
@nsinghal12 are you sure you added the new yarn.lock and pushed it? GitHub is still showing it as deleted.

**Comment by crenshaw-dev** [comment]
Oh, I see. Make sure you run `yarn install` from inside the ui directory.

**Comment by blakepettersson** [comment]
>  I think you meant yarn.lock file.

@nsinghal12 yes, that was a silly typo 😄 

As @crenshaw-dev said, the `yarn.lock` should be regenerated by running `yarn install` in the `ui/` dir.

**Comment by crenshaw-dev** [comment]
Closing for now as this needs more work.

**Pull request closed without merging**